### PR TITLE
[WIP] adjoint for steady state solution

### PR DESCRIPTION
**Pull request by frankschae**
I tried to implement the adjoint sensitivity function for the steady state solution #182 . I didn't manage to make the linsolve work, yet .. but I assumed we'd want that a user can choose the linear solver, too.. 

Also, I was not sure if del g/ del p in (3) of https://math.mit.edu/~stevenj/18.336/adjoint.pdf should be added automatically by AD/FD or by a user input. (For the tests in adjoint.jl I chose a loss function that doesn't explicitly depend on p so far. )

### CODE REVIEW

File: test/adjoint.jl [hunk 1]
```diff
@@ -660,3 +660,136 @@ using LinearAlgebra, DiffEqSensitivity, OrdinaryDiffEq, ForwardDiff, QuadGK
     end
   end
 end
+
+
+using Test, LinearAlgebra
```
Comment (ChrisRackauckas) [inline test/adjoint.jl:665]: lets make this a different file.

File: src/local_sensitivity/sensitivity_interface.jl [hunk 2]
```diff
@@ -22,6 +22,19 @@ function _adjoint_sensitivities(sol,sensealg,alg,g,t=nothing,dg=nothing;
...
 
+function _adjoint_sensitivities(sol,sensealg::SteadyStateAdjoint,alg,g,dg=nothing;
+                                   abstol=1e-6,reltol=1e-3,
+                                   kwargs...)
+  cache = SteadyStateAdjointSensitivityFunction(g,sensealg,sol,dg)
+
+  linsolve_tmp = cache.J'\vec(cache.gu')
+  #cache.linsolve(vec(cache.gu), cache.J', vec(cache.linsolve_tmp))
+
+  -linsolve_tmp'*cache.pJ #dg/dp?
```
Comment (ChrisRackauckas) [inline src/local_sensitivity/sensitivity_interface.jl:33]: yeah only the discrete form really works here so we should be taking dg/dp. Concrete solve might even be easier since that's the dy value from the backpass.

File: src/local_sensitivity/sensitivity_interface.jl [hunk 3]
```diff
@@ -22,6 +22,19 @@ function _adjoint_sensitivities(sol,sensealg,alg,g,t=nothing,dg=nothing;
     adj_sol[end][(1:l) .+ length(sol.prob.u0)]'
 end
 
+function _adjoint_sensitivities(sol,sensealg::SteadyStateAdjoint,alg,g,dg=nothing;
+                                   abstol=1e-6,reltol=1e-3,
+                                   kwargs...)
+  cache = SteadyStateAdjointSensitivityFunction(g,sensealg,sol,dg)
+
+  linsolve_tmp = cache.J'\vec(cache.gu')
```
Comment (ChrisRackauckas) [inline src/local_sensitivity/sensitivity_interface.jl:30]: all that should we need to do is compute the vector'*Jacobian, which is the vecjacobian! stuff of the other adjoints. Copy that general workflow, since we want to use backprop on f for that. You might want to read some of the 18.337 notes, specifically https://mitmath.github.io/18337/lecture10/estimation_identification on vjps

Comment (frankschae) [reply src/local_sensitivity/sensitivity_interface.jl:30]: Thanks for the remarks and the link to the notes! I'll read them and try to update the files accordingly.

**Comment by ChrisRackauckas** [comment]
I think this is probably good. Follow up PRs can handle the extra case and linsolve, but I think it would be good to the basics cleaned up and done first.

The other thing to handle will be `concrete_solve`, but that shouldn't be too difficult given what I see here.

**Comment by frankschae** [comment]
Is there something missing before we can merge it? I am planning to do the `concrete_solve` next .
